File: js/1441.build-an-array-with-stack-operations.js
/*
 * @lc app=leetcode id=1441 lang=javascript
 *
 * [1441] Build an Array With Stack Operations
 */

// @lc code=start
/**
 * @param {number[]} target
 * @param {number} n
 * @return {string[]}
 */
var buildArray = function(target, n) {
  let output = []
  let diff = target[0] - 1
  output = addPushPop(output, diff)

  for (let i=1; i<target.length; i++) {
    diff = target[i] - target[i-1] - 1
    output = addPushPop(output, diff)
  }
  return output
};

let addPushPop = (arr, n) => {
  let add = ',Push,Pop'.repeat(n).slice(1).split(',')
  if (!add[0].length) add = []
  add.push('Push')
  return [...arr, ...add]
}
// @lc code=end

File: 1441.build-an-array-with-stack-operations.py
#
# @lc app=leetcode id=1441 lang=python3
#
# [1441] Build an Array With Stack Operations
#

# @lc code=start
class Solution:
    def buildArray(self, target: List[int], n: int) -> List[str]:
        iTarget = 0
        ops = []
        j=0
        while j < n and iTarget < len(target):
            num = target[iTarget]
            lastNum = target[iTarget-1] if iTarget > 0 else 0
            numDiff = num - lastNum
            for _ in range(1, numDiff):
                ops.append('Push')
                ops.append('Pop')
            ops.append('Push')
            iTarget += 1
            j += 1

        return ops
# @lc code=end

